datapath: Fix update to NodeAddress table and prioritize Node IP

joamaki opened the pull request:
When new node addresses were added after Cilium had started the update to the internal table of node addresses accidentally removed existing addresses due to wrong kind of comparison operation.

As part of fixing this issue a similar type of problem was found with update to the fallback addresses (used as backup when selecting BPF masquerading address). This commit also fixes this issue.

This only impacts users running v1.15 with the runtime device detection enabled.

The second commit adds in the prioritization of the K8s Node IP when selecting the NodeAddress. This was a semantic
difference between the old and new code and should have been retained.

Fixes: #33234 

```release-note
Fix an issue in updates to node addresses which may have caused missing NodePort frontend IP addresses. May have affected NodePort/LoadBalancer services for users running with runtime device detection enabled when node's IP addresses were changed after Cilium had started.
Node IP as defined in the Kubernetes Node is now preferred when selecting the NodePort frontend IPs.
```

joamaki committed:
datapath: Fix incremental update to NodeAddress table

When new node addresses were added after Cilium had started the
update to the internal table of node addresses accidentally removed
existing addresses due to comparing the NodeAddress instead of netip.Addr
which caused change in e.g. the NodePort flag to indicate that the address
should be removed.

As an example:
  Before:
    NodeAddress{Device: "eth0", Addr: "10.0.0.5", NodePort: true}

  After (expected):
    NodeAddress{Device: "eth0", Addr: "10.0.0.1", NodePort: true}
    NodeAddress{Device: "eth0", Addr: "10.0.0.5", NodePort: false}

  After (actual):
    NodeAddress{Device: "eth0", Addr: "10.0.0.1", NodePort: true}

What happened was that the code was removing anything old that wasn't in
the new set of addresses, leading it to essentially this comparison:
    NodeAddress{"eth0", "10.0.0.5", true} !=
	  NodeAddress{"eth0", "10.0.0.5", false}

The result of which was that the address "10.0.0.5" was updated and then
subsequently deleted.

Fix this by reworking the code to keep the old set of addresses as netip.Addr
and reducing that set down when inserting the new entries and then finally
cleaning up what remains.

The prefix match comment on the device name and the prefix length check was
removed as this issue had been fixed in StateDB.

As far as we can tell, the only production impact this could have had would
have been to users of v1.16 pre-releases or users running v1.15 with the
experimental runtime device detection enabled (only then would have changes
to Table[NodeAddress] mattered).

Fixes: #33234
Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
datapath: Prefer node IPs for NodePort and Primary

The K8s Node IP address should be preferred when choosing the NodePort and
Primary (BPF masquerade) addresses.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

dylandreimerink commented on pkg/datapath/tables/node_address.go:
```diff
@@ -436,12 +476,13 @@ func (n *nodeAddressController) getAddressesFromDevice(dev *Device) sets.Set[Nod
-			if !ipv4Found {
-				ipv4Found = true
-				primary = true
+			if addr.Addr.Unmap() == n.k8sIPv4.Unmap() {
```
The `Unmap` seems unnecessary if both IPs are confirmed to be IPv4 addresses

joamaki replied:
What would you propose doing instead? If I remove Unmap the tests fail as one of these ends up being an IPv4-mapped IPv6 address (either MustParseAddr or AddrFromSlice causes it, probably the latter...).

EDIT: could use our `ip.AddrFromIP` instead, but perhaps safer to compare with Unmap'd addresses still?

dylandreimerink replied:
> If I remove Unmap the tests fail as one of these ends up being an IPv4-mapped IPv6 address

That is odd, the branch is guarded by an `if addr.Addr.Is4() {` check so `addr.Addr` should be an IPv4 for sure. Which means that we allow IPv4-mapped-IPv6 addresses as valid value for `n.k8sIPv4`.

If we truly support that then there is nothing wrong with the code. I just didn't expected that

dylandreimerink reviewed (approved):
Just a small nit

joamaki committed:
datapath: Fix update of fallback addresses when address is removed

The fallback addresses in Table[NodeAddress] were not properly updated
when the address was removed. Fix this by clearing the addresses first
when updating the fallbacks.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

brb reviewed (changes requested):
As discussed offline, let's split the first commit into two, and then update the commit msg to include more info on the first fix.

joamaki commented on pkg/datapath/tables/node_address.go:
```diff
@@ -265,24 +272,43 @@ func (n *nodeAddressController) register() {
 		for change, _, ok := n.deviceChanges.Next(); ok; change, _, ok = n.deviceChanges.Next() {
 			dev := change.Object
 
-			// Note: prefix match! existing may contain node addresses from devices with names
```
Note to self/backporter: Verify whether this code is needed in v1.15 or whether this workaround is still needed.

brb reviewed (approved):
Thanks!
